DataScience/experimenter.py

```diff
@@ -0,0 +1,202 @@
+        experiment_file.write(line + "\n")
+        experiment_file.flush()
+        
+    experiment_file.close()
+    
+def run_experiment(command, timeout=1000):
+    while True:
+        try:
+            results = check_output(command.full_command.split(' '), stderr=STDOUT, timeout=timeout).decode("utf-8")
+            break
+        except TimeoutExpired:
+            print("Timeout expired for command {0}".format(command.full_command))
+            
+    m = re.search('average loss = (.+)\n', str(results))
+    loss = m.group(1)
+    command.loss = float(loss)
+    result_queue.put(command)
+    return command
+    
+def run_experiment_set(command_list):
+    # Run the experiments in parallel using 5 processes
+    p = Pool(5)
+    results = p.map(run_experiment, command_list)
+    results.sort(key=lambda result: result.loss)
+    return results
+    
+if __name__ == '__main__':
+    # Identify namespaces and detect marginal features
+    if len(sys.argv) < 2:
+        print("Usage: python experimenter.py {file_name}. Where file_name is the merged Decision Service logs in JSON format")
+        sys.exit()
+    data = open(sys.argv[1], 'r', encoding="utf8")
+    counter = 0
+    shared_features = []
+    action_features = []
+    marginal_features = []
+    for line in data:
+        counter += 1
+        event = json.loads(line)
+        for feature in event.keys():
+            if feature[0] != '_' and feature[0] not in shared_features:
+                shared_features.append(feature[0])
+        action_set = event['_multi']
+        for action in action_set:
+            for feature in action.keys():
+                if feature[0] != '_' and feature[0] not in action_features:
+                    action_features.append(feature[0])
+                    if action[feature].get('constant', 0) == 1 and 'id' in action[feature]:
+                        marginal_features.append(feature[0])
+        if counter >= 50:
+            break
+            
+    data.close()
+    print("Shared feature namespaces: " + str(shared_features))
+    print("Action feature namespaces: " + str(action_features))
+    print("Marginal feature namespaces: " + str(marginal_features))
+
+    # Start the writing process for experiment results
+    result_queue = Queue()
+    writer_p = Process(target=result_writer, args=((result_queue),))
+    writer_p.daemon = True
+    writer_p.start()        
+    
+    # Base parameters and setting up the cache file
+    base_command = "vw --cb_adf -d %s --json -c --power_t 0" % sys.argv[1] # TODO: VW location should be a command line parameter. 
+    initial_command = Command(base_command, learning_rate=0.5)
+    run_experiment(initial_command, timeout=3600)
+    
+    # Learning rates
+    command_list = []
+    learning_rates = [0.5, 0.25, 0.1, 0.05, 0.025, 0.01, 0.005, 0.0025, 0.001, 0.0005, 0.00025, 0.0001, 0.00005, 0.000025, 0.00001]
```

Comment:
externalize to ini/config file? the ini module I used for Azure download was straight to use